Implement z_mergetoaddress for combining UTXOs and notes

str4d opened the pull request:
Closes #2493.

str4d commented:
I'm still working on the RPC test, but I've opened this PR to start garnering review on the code.

tromer commented:
Nice! Also closes #1785.

How does this handle coinbase?

str4d commented:
That's a good point - currently it will try to include them, and thus fail if the target is a t-addr. Given that we have `z_shieldcoinbase`, I propose that this method simply ignore coinbase UTXOs.

bitcartel commented:
Let's make this an experimental feature, just as z_shieldcoinbase was at first.

str4d commented:
Bumping to 1.0.15 due to necessary work.

bitcartel commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3707,6 +3708,302 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+            "    ]\n"
+            "2. \"toaddress\"           (string, required) The taddr or zaddr to send the funds to.\n"
+            "3. fee                   (numeric, optional, default="
+            + strprintf("%s", FormatMoney(SHIELD_COINBASE_DEFAULT_MINERS_FEE)) + ") The fee amount to attach to this transaction.\n"
```
Should be MERGETOADDRESS_OPERATION_DEFAULT_MINERS_FEE

bitcartel commented on src/wallet/asyncrpcoperation_mergetoaddress.h:
```diff
@@ -0,0 +1,207 @@
+#include <univalue.h>
+
+// Default transaction fee if caller does not specify one.
+#define MERGETOADDRESS_OPERATION_DEFAULT_MINERS_FEE 10000
```
Or MERGE_TO_ADDRESS_... to match formatting of other M_T_A_... constants.

str4d commented:
Rebased on master to fix merge conflicts, added unit tests, finished RPC test, fixed bugs and missing details. I'll add another commit shortly that gates `z_mergetoaddress` under experimental features.

bitcartel commented on src/rpcclient.cpp:
```diff
@@ -109,6 +109,10 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "z_gettotalbalance", 0},
     { "z_gettotalbalance", 1},
     { "z_gettotalbalance", 2},
+    { "z_mergetoaddress", 0},
```
Remove `{ "z_mergetoaddress", 0},`.  Causes bug where argument like "*" causes JSON parsing errors (later in review).

str4d replied:
Yeah, it's annoying. I think the way to fix this properly is to remove this line, and then if the string doesn't match one of the expected constants, manually call the JSON parser.

str4d replied:
Follow-up question: Why does the RPC test I wrote not fail? It uses both string and array arguments...

str4d replied:
Ah, it's because this line affects how `zcash-cli` parses the CLI inputs. This is kind of annoying, because it means that if we remove this line, `zcash-cli` sends an unparsed JSON string representation as an actual string.

Okay, to fix this, I'm going to just make it always an array, but it can contain the keywords.

bitcartel commented on src/wallet/asyncrpcoperation_mergetoaddress.h:
```diff
@@ -0,0 +1,187 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef ASYNCRPCOPERATION_MERGETOADDRESS_H
```
Attaching comment here about a reproducbile anchor problem when merging using "*" when there are both  utxos and notes present:

```
./zcash-cli z_mergetoaddress "*" ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 0.0001 0 0 "BEEF"
{
  "remainingUTXOs": 0,
  "remainingTransparentValue": 0.00000000,
  "remainingNotes": 0,
  "remainingShieldedValue": 0.00000000,
  "mergingUTXOs": 2,
  "mergingTransparentValue": 200.00000000,
  "mergingNotes": 1,
  "mergingShieldedValue": 1575.31036597,
  "opid": "opid-8e1ac86c-dc35-4465-b529-9a8fbd849aea"
}
```
This fails:
```
2018-01-25 21:17:48 opid-8e1ac86c-dc35-4465-b529-9a8fbd849aea: z_mergetoaddress initialized (params={"fromaddresses":"*","toaddress":"ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE","fee":0.00010000})
2018-01-25 21:17:48 opid-8e1ac86c-dc35-4465-b529-9a8fbd849aea: spending 1775.31036597 to send 1775.31026597 with fee 0.0001
2018-01-25 21:17:48 opid-8e1ac86c-dc35-4465-b529-9a8fbd849aea: transparent input: 200.00
2018-01-25 21:17:48 opid-8e1ac86c-dc35-4465-b529-9a8fbd849aea: private input: 1575.31036597
2018-01-25 21:17:48 opid-8e1ac86c-dc35-4465-b529-9a8fbd849aea: private output: 1775.31026597
2018-01-25 21:17:48 opid-8e1ac86c-dc35-4465-b529-9a8fbd849aea: fee: 0.0001
2018-01-25 21:17:48 opid-8e1ac86c-dc35-4465-b529-9a8fbd849aea: z_mergetoaddress finished (status=failed, error=Could not find previous JoinSplit anchor)
```
However, calling "ANY_ZADDR" at this point succeeds, as does "ANY_ZADDR".
Here is the "ANY_ZADDR" output:
```
./zcash-cli z_mergetoaddress "ANY_ZADDR" ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 0.0001 0 0 "BEEF"
{
  "remainingUTXOs": 0,
  "remainingTransparentValue": 0.00000000,
  "remainingNotes": 0,
  "remainingShieldedValue": 0.00000000,
  "mergingUTXOs": 0,
  "mergingTransparentValue": 0.00000000,
  "mergingNotes": 1,
  "mergingShieldedValue": 1575.31036597,
  "opid": "opid-deec2331-c1c2-4a75-9314-b72808b95255"
}
```
and log output:
```
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: z_mergetoaddress initialized (params={"fromaddresses":"ANY_ZADDR","toaddress":"ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE","fee":0.00010000})
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: spending 1575.31036597 to send 1575.31026597 with fee 0.0001
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: transparent input: 0.00
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: private input: 1575.31036597
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: private output: 1575.31026597
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: fee: 0.0001
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: spending note (txid=1ae2ac96d5, vjoinsplit=0, ciphertext=0, amount=1575.31036597, height=652, confirmations=1)
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: generating note for change (amount=1575.31026597)
2018-01-25 21:22:10 opid-deec2331-c1c2-4a75-9314-b72808b95255: creating joinsplit at index 0 (vpub_old=0.00, vpub_new=0.0001, in[0]=1575.31036597, in[1]=0.00, out[0]=0.00, out[1]=1575.31026597)
2018-01-25 21:22:46 opid-deec2331-c1c2-4a75-9314-b72808b95255: Payment Disclosure: js=0, n=1, zaddr=ztkjJKs99AqHhoUUHa5h5cJ44tmQvBqgk66D2unWhk1xkP2L36SxhufEQLj5WyeS9uy8c5A4ZvpfEWuQkHE6PEdEuzUPkqt
2018-01-25 21:22:46 opid-deec2331-c1c2-4a75-9314-b72808b95255: Payment Disclosure: js=0, n=0, zaddr=ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE
2018-01-25 21:22:46 Blockpolicy mempool tx 05b61e6592 not adding
2018-01-25 21:22:46 AddToWallet 05b61e6592fb92a4cd4497a2730009737fd2f9899974c66a195a2368c58edcae  new
2018-01-25 21:22:46 opid-deec2331-c1c2-4a75-9314-b72808b95255: z_mergetoaddress finished (status=success, txid=05b61e6592fb92a4cd4497a2730009737fd2f9899974c66a195a2368c58edcae)
```

bitcartel replied:
The above is also reproducible when manually specifying the array of addresses e.g. '["t....", "z....."]'

bitcartel commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3726,6 +3727,317 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+            "                         The following special strings are also accepted:\n"
+            "                             - \"*\": Merge both UTXOs and notes from all addressed belonging to the wallet.\n"
+            "                             - \"ANY_TADDR\": Merge UTXOs from all taddrs belonging to the wallet.\n"
+            "                             - \"ANY_ZADDR\": Merge notes from all zaddrs belonging to the wallet.\n"
```
Fixed by my comment in `src/rpcclient.cpp`

bitcartel commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3726,6 +3727,317 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+            "\nArguments:\n"
+            "1. fromaddresses         (string, required) A JSON array with addresses.\n"
+            "                         The following special strings are also accepted:\n"
+            "                             - \"*\": Merge both UTXOs and notes from all addressed belonging to the wallet.\n"
```
Fixed by my comment in `src/rpcclient.cpp`

bitcartel commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3726,6 +3727,317 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+            "1. fromaddresses         (string, required) A JSON array with addresses.\n"
+            "                         The following special strings are also accepted:\n"
+            "                             - \"*\": Merge both UTXOs and notes from all addressed belonging to the wallet.\n"
+            "                             - \"ANY_TADDR\": Merge UTXOs from all taddrs belonging to the wallet.\n"
```
Tested, yes this is case sensitive.  Should we make this case insensitive?

str4d replied:
Simple enough to do so, if we think it is beneficial.

bitcartel commented on src/wallet/asyncrpcoperation_mergetoaddress.h:
```diff
@@ -0,0 +1,187 @@
+#ifndef ASYNCRPCOPERATION_MERGETOADDRESS_H
+#define ASYNCRPCOPERATION_MERGETOADDRESS_H
+
+#include "amount.h"
```
Attaching here as comment:

Funds were merged into a different zaddr (in the wallet), and not the zaddr (in the wallet) specified.  This was run on mainnet.

```
./zcash-cli z_mergetoaddress ANY_ZADDR zcdYqZzp5ABTtsKi8PdEZefwjWfbeuBMNRn91pWcLsbZAztDcUBXwx25aaGk1NoSCUP66AA5VMBhv9kCQE3V1zTMYUFRERy 0.00001
{
  "remainingUTXOs": 0,
  "remainingTransparentValue": 0.00000000,
  "remainingNotes": 0,
  "remainingShieldedValue": 0.00000000,
  "mergingUTXOs": 0,
  "mergingTransparentValue": 0.00000000,
  "mergingNotes": 1,
  "mergingShieldedValue": 0.03872239,
  "opid": "opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b"
}
```

debug.log:

```
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: z_mergetoaddress initialized (params={"fromaddresses":"ANY_ZADDR","toaddress":"zcdYqZzp5ABTtsKi8PdEZefwjWfbeuBMNRn91pWcLsbZAztDcUBXwx25aaGk1NoSCUP66AA5VMBhv9kCQE3V1zTMYUFRERy","fee":0.00001000})
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: spending 0.03872239 to send 0.03871239 with fee 0.00001
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: transparent input: 0.00
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: private input: 0.03872239
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: private output: 0.03871239
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: fee: 0.00001
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: spending note (txid=1404b71a86, vjoinsplit=0, ciphertext=0, amount=0.03872239, height=260958, confirmations=10)
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: generating note for change (amount=0.03871239)
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: creating joinsplit at index 0 (vpub_old=0.00, vpub_new=0.00001, in[0]=0.03872239, in[1]=0.00, out[0]=0.00, out[1]=0.03871239)
...
2018-01-26 00:55:35 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: Payment Disclosure: js=0, n=1, zaddr=zcDEb4i1nYPHgAABg86TuhoBLJ1aSTpA8kK4bE32eXcjaJdABKymaG5YAWfgiyGCk9mZocXQFHjBeuGkBFpAGqAv5ERDtYx
2018-01-26 00:55:35 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: Payment Disclosure: js=0, n=0, zaddr=zcUxZKkUd6TwWtxYxJta2A4kYkz4qHoyEdJBiB4FvyKTheYdjHNNJUR73qMmFZw4rqwbTjwLxjhYMBkgwQxVWKPqawmULqp
...
2018-01-26 00:55:35 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: z_mergetoaddress finished (status=success, txid=0123d0c0a003cabd3eb7460a73e78d258f0d665c6904d224100d25249cd1f12e)
```

Payment disclosure shows index 0 using an address in the wallet, whereas index 1 is a dummy address.

```
Result is that funds found their way to a different zaddr in wallet, as shown by index 0 of js.  

./zcash-cli z_getbalance zcUxZKkUd6TwWtxYxJta2A4kYkz4qHoyEdJBiB4FvyKTheYdjHNNJUR73qMmFZw4rqwbTjwLxjhYMBkgwQxVWKPqawmULqp
0.03871239
```

The test above was to an address in the same wallet.  Have not tried with an address in a different wallet.

str4d replied:
Added a case to the RPC test and replicated the bug. Seems like the to address is being ignored in this case, and all funds are being returned to the change address (see the `generating note for change` line in the trace).

str4d replied:
Simple bug - I had forgotten to alter the target address of the final change output, so it was going to the z-address for the first note (which gets used as the change address for intermediate change).

bitcartel commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3726,6 +3727,317 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+    std::set<CBitcoinAddress> taddrs = {};
+    std::set<libzcash::PaymentAddress> zaddrs = {};
+    if (!(useAny || useAnyUTXO || useAnyNote)) {
+        UniValue outputs = params[0].get_array();
```
Variable name --> inputs

str4d replied:
Changing to `addresses`.

bitcartel commented on src/wallet/wallet.h:
```diff
@@ -1126,6 +1127,13 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
                           bool ignoreUnspendable=true);
+
+    /* Find notes filtered by payment addresses, min depth, ability to spend */
+    void GetFilteredNotes(std::vector<CNotePlaintextEntry>& outEntries,
```
Related: See #2913 where I added a Get.. method for z_listunspent.  Looked to refactor, but it needed a maxDepth parameter and also a way to pass back the number of confirmations, which doesn't really belong to struct CNotePlaintextEntry.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3726,6 +3727,317 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+
+#define JOINSPLIT_SIZE JSDescription().GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION)
+
+UniValue z_mergetoaddress(const UniValue& params, bool fHelp)
```
That's an excellent point! 😆

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3726,6 +3727,317 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+            + HelpRequiringPassphrase() + "\n"
+            "\nArguments:\n"
+            "1. fromaddresses         (string, required) A JSON array with addresses.\n"
+            "                         The following special strings are also accepted:\n"
```
I'll clarify the help text.

str4d replied:
Clarified to explain that the above is now actually allowed 😄

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3726,6 +3727,317 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+
+    if (fHelp || params.size() < 2 || params.size() > 6)
+        throw runtime_error(
+            "z_mergetoaddress [\"fromaddress\", ... ] \"toaddress\" ( fee ) ( transparent_limit ) ( shielded_limit ) ( memo )\n"
```
Fixed to use `transparent_limit` and `shielded_limit`. Re: UTXO and note usage, I figure it makes sense when talking about specific objects, but we should probably do a review of the whole RPC API's help text for consistency at some point.

str4d commented:
Pushed new commits to address most of @bitcartel's comments. I will squash them all together after review. I still need to reproduce the anchor bug.

str4d commented:
Addressed all of @bitcartel's comments.

str4d commented:
Rebased on master and adjusted the PR to account for the Overwinter transaction format and SignatureHash. Diff below:

```diff
diff --git a/src/test/rpc_wallet_tests.cpp b/src/test/rpc_wallet_tests.cpp
index f3b95ef..18b687d 100644
--- a/src/test/rpc_wallet_tests.cpp
+++ b/src/test/rpc_wallet_tests.cpp
@@ -1487,6 +1487,11 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_parameters)
     BOOST_CHECK_THROW(CallRPC(string("z_mergetoaddress [\"tmRr6yJonqGK23UVhrKuyvTpF8qxQQjKigJ\"] ")
             + zaddr1 + " 0.0001 100 100 " + badmemo), runtime_error);
 
+    // Mutable tx containing contextual information we need to build tx
+    UniValue retValue = CallRPC("getblockcount");
+    int nHeight = retValue.get_int();
+    CMutableTransaction mtx = CreateNewContextualCMutableTransaction(Params().GetConsensus(), nHeight + 1);
+
     // Test constructor of AsyncRPCOperation_mergetoaddress
     MergeToAddressRecipient testnetzaddr(
         "ztjiDe569DPNbyTE6TSdJTaSDhoXEHLGvYoUnBU1wfVNU52TEyT6berYtySkd21njAeEoh8fFJUT42kua9r8EnhBaEKqCpP",
@@ -1496,14 +1501,14 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_parameters)
         "mainnet memo");
 
     try {
-        std::shared_ptr<AsyncRPCOperation> operation(new AsyncRPCOperation_mergetoaddress({}, {}, testnetzaddr, -1 ));
+        std::shared_ptr<AsyncRPCOperation> operation(new AsyncRPCOperation_mergetoaddress(mtx, {}, {}, testnetzaddr, -1 ));
         BOOST_FAIL("Should have caused an error");
     } catch (const UniValue& objError) {
         BOOST_CHECK( find_error(objError, "Fee is out of range"));
     }
 
     try {
-        std::shared_ptr<AsyncRPCOperation> operation(new AsyncRPCOperation_mergetoaddress({}, {}, testnetzaddr, 1));
+        std::shared_ptr<AsyncRPCOperation> operation(new AsyncRPCOperation_mergetoaddress(mtx, {}, {}, testnetzaddr, 1));
         BOOST_FAIL("Should have caused an error");
     } catch (const UniValue& objError) {
         BOOST_CHECK( find_error(objError, "No inputs"));
@@ -1513,7 +1518,7 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_parameters)
 
     try {
         MergeToAddressRecipient badaddr("", "memo");
-        std::shared_ptr<AsyncRPCOperation> operation(new AsyncRPCOperation_mergetoaddress(inputs, {}, badaddr, 1));
+        std::shared_ptr<AsyncRPCOperation> operation(new AsyncRPCOperation_mergetoaddress(mtx, inputs, {}, badaddr, 1));
         BOOST_FAIL("Should have caused an error");
     } catch (const UniValue& objError) {
         BOOST_CHECK( find_error(objError, "Recipient parameter missing"));
@@ -1522,7 +1527,7 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_parameters)
     // Testnet payment addresses begin with 'zt'.  This test detects an incorrect prefix.
     try {
         std::vector<MergeToAddressInputUTXO> inputs = { MergeToAddressInputUTXO{ COutPoint{uint256(), 0}, 0} };
-        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(inputs, {}, mainnetzaddr, 1) );
+        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(mtx, inputs, {}, mainnetzaddr, 1) );
         BOOST_FAIL("Should have caused an error");
     } catch (const UniValue& objError) {
         BOOST_CHECK( find_error(objError, "payment address is for wrong network type"));
@@ -1537,7 +1542,10 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_internals)
 
     LOCK(pwalletMain->cs_wallet);
 
-    UniValue retValue;
+    // Mutable tx containing contextual information we need to build tx
+    UniValue retValue = CallRPC("getblockcount");
+    int nHeight = retValue.get_int();
+    CMutableTransaction mtx = CreateNewContextualCMutableTransaction(Params().GetConsensus(), nHeight + 1);
 
     // Test that option -mempooltxinputlimit is respected.
     mapArgs["-mempooltxinputlimit"] = "1";
@@ -1554,7 +1562,7 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_internals)
             MergeToAddressInputUTXO{COutPoint{uint256(),0},0},
             MergeToAddressInputUTXO{COutPoint{uint256(),0},0}
         };
-        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(inputs, {}, zaddr1) );
+        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(mtx, inputs, {}, zaddr1) );
         operation->main();
         BOOST_CHECK(operation->isFailed());
         std::string msg = operation->getErrorMessage();
@@ -1564,7 +1572,7 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_internals)
     // Insufficient funds
     {
         std::vector<MergeToAddressInputUTXO> inputs = { MergeToAddressInputUTXO{COutPoint{uint256(),0},0} };
-        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(inputs, {}, zaddr1) );
+        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(mtx, inputs, {}, zaddr1) );
         operation->main();
         BOOST_CHECK(operation->isFailed());
         std::string msg = operation->getErrorMessage();
@@ -1574,7 +1582,7 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_internals)
     // get_memo_from_hex_string())
     {
         std::vector<MergeToAddressInputUTXO> inputs = { MergeToAddressInputUTXO{COutPoint{uint256(),0},100000} };
-        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(inputs, {}, zaddr1) );
+        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(mtx, inputs, {}, zaddr1) );
         std::shared_ptr<AsyncRPCOperation_mergetoaddress> ptr = std::dynamic_pointer_cast<AsyncRPCOperation_mergetoaddress> (operation);
         TEST_FRIEND_AsyncRPCOperation_mergetoaddress proxy(ptr);
 
@@ -1628,7 +1636,7 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_internals)
     {
         // Dummy input so the operation object can be instantiated.
         std::vector<MergeToAddressInputUTXO> inputs = { MergeToAddressInputUTXO{COutPoint{uint256(),0},100000} };
-        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(inputs, {}, zaddr1) );
+        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(mtx, inputs, {}, zaddr1) );
         std::shared_ptr<AsyncRPCOperation_mergetoaddress> ptr = std::dynamic_pointer_cast<AsyncRPCOperation_mergetoaddress> (operation);
         TEST_FRIEND_AsyncRPCOperation_mergetoaddress proxy(ptr);
 
@@ -1689,7 +1697,7 @@ BOOST_AUTO_TEST_CASE(rpc_z_mergetoaddress_internals)
 
         // we have the spending key for the dummy recipient zaddr1
         std::vector<MergeToAddressInputUTXO> inputs = { MergeToAddressInputUTXO{COutPoint{uint256(),0},100000} };
-        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(inputs, {}, zaddr1) );
+        std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_mergetoaddress(mtx, inputs, {}, zaddr1) );
         std::shared_ptr<AsyncRPCOperation_mergetoaddress> ptr = std::dynamic_pointer_cast<AsyncRPCOperation_mergetoaddress> (operation);
         TEST_FRIEND_AsyncRPCOperation_mergetoaddress proxy(ptr);
 
diff --git a/src/wallet/asyncrpcoperation_mergetoaddress.cpp b/src/wallet/asyncrpcoperation_mergetoaddress.cpp
index de3c546..823f289 100644
--- a/src/wallet/asyncrpcoperation_mergetoaddress.cpp
+++ b/src/wallet/asyncrpcoperation_mergetoaddress.cpp
@@ -52,11 +52,14 @@ int mta_find_output(UniValue obj, int n)
 }
 
 AsyncRPCOperation_mergetoaddress::AsyncRPCOperation_mergetoaddress(
+    CMutableTransaction contextualTx,
     std::vector<MergeToAddressInputUTXO> utxoInputs,
     std::vector<MergeToAddressInputNote> noteInputs,
     MergeToAddressRecipient recipient,
     CAmount fee,
-    UniValue contextInfo) : utxoInputs_(utxoInputs), noteInputs_(noteInputs), recipient_(recipient), fee_(fee), contextinfo_(contextInfo)
+    UniValue contextInfo) :
+    tx_(contextualTx), utxoInputs_(utxoInputs), noteInputs_(noteInputs),
+    recipient_(recipient), fee_(fee), contextinfo_(contextInfo)
 {
     if (fee < 0 || fee > MAX_MONEY) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Fee is out of range");
@@ -251,6 +254,12 @@ bool AsyncRPCOperation_mergetoaddress::main_impl()
     }
     LogPrint("zrpc", "%s: fee: %s\n", getId(), FormatMoney(minersFee));
 
+    // Grab the current consensus branch ID
+    {
+        LOCK(cs_main);
+        consensusBranchId_ = CurrentEpochBranchId(chainActive.Height() + 1, Params().GetConsensus());
+    }
+
     /**
      * SCENARIO #1
      *
@@ -271,7 +280,9 @@ bool AsyncRPCOperation_mergetoaddress::main_impl()
 
     // Prepare raw transaction to handle JoinSplits
     CMutableTransaction mtx(tx_);
-    mtx.nVersion = 2;
+    if (!mtx.fOverwintered) {
+        mtx.nVersion = 2;
+    }
     crypto_sign_keypair(joinSplitPubKey_.begin(), joinSplitPrivKey_);
     mtx.joinSplitPubKey = joinSplitPubKey_;
     tx_ = CTransaction(mtx);
@@ -771,7 +782,7 @@ UniValue AsyncRPCOperation_mergetoaddress::perform_joinsplit(
     // Empty output script.
     CScript scriptCode;
     CTransaction signTx(mtx);
-    uint256 dataToBeSigned = SignatureHash(scriptCode, signTx, NOT_AN_INPUT, SIGHASH_ALL);
+    uint256 dataToBeSigned = SignatureHash(scriptCode, signTx, NOT_AN_INPUT, SIGHASH_ALL, 0, consensusBranchId_);
 
     // Add the signature
     if (!(crypto_sign_detached(&mtx.joinSplitSig[0], NULL,
diff --git a/src/wallet/asyncrpcoperation_mergetoaddress.h b/src/wallet/asyncrpcoperation_mergetoaddress.h
index d8fb0fa..1619b5c 100644
--- a/src/wallet/asyncrpcoperation_mergetoaddress.h
+++ b/src/wallet/asyncrpcoperation_mergetoaddress.h
@@ -53,6 +53,7 @@ class AsyncRPCOperation_mergetoaddress : public AsyncRPCOperation
 {
 public:
     AsyncRPCOperation_mergetoaddress(
+        CMutableTransaction contextualTx,
         std::vector<MergeToAddressInputUTXO> utxoInputs,
         std::vector<MergeToAddressInputNote> noteInputs,
         MergeToAddressRecipient recipient,
@@ -79,6 +80,7 @@ private:
 
     UniValue contextinfo_; // optional data to include in return value from getStatus()
 
+    uint32_t consensusBranchId_;
     CAmount fee_;
     int mindepth_;
     MergeToAddressRecipient recipient_;
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index 23a91e9..17bdc72 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -4056,10 +4056,15 @@ UniValue z_mergetoaddress(const UniValue& params, bool fHelp)
     contextInfo.push_back(Pair("toaddress", params[1]));
     contextInfo.push_back(Pair("fee", ValueFromAmount(nFee)));
 
+    // Contextual transaction we will build on
+    CMutableTransaction contextualTx = CreateNewContextualCMutableTransaction(
+        Params().GetConsensus(),
+        chainActive.Height() + 1);
+
     // Create operation and add to global queue
     std::shared_ptr<AsyncRPCQueue> q = getAsyncRPCQueue();
     std::shared_ptr<AsyncRPCOperation> operation(
-        new AsyncRPCOperation_mergetoaddress(utxoInputs, noteInputs, recipient, nFee, contextInfo) );
+        new AsyncRPCOperation_mergetoaddress(contextualTx, utxoInputs, noteInputs, recipient, nFee, contextInfo) );
     q->addOperation(operation);
     AsyncRPCOperationId operationId = operation->getId();
 ```

bitcartel commented:
Will perform live tests tomorrow.

arielgabizon commented:
I'm not managing to use the "*" option  
![image](https://user-images.githubusercontent.com/5689834/36483343-85b8b592-171e-11e8-8da8-1605021df2f9.png)

str4d commented:
@arielgabizon in the CLi you need to quote JSON inputs, so use either `z_mergetoaddress '["*"]' ...` or `z_mergetoaddress "[\"*\"]" ...`

arielgabizon commented:
OK, just turned out I needed external '' brackets (that aren't in the examples btw)
Now having another issue - says I don't have notes to merge though have many utxos
![image](https://user-images.githubusercontent.com/5689834/36487006-92e73112-1728-11e8-88e6-ee1f2210a62d.png)

arielgabizon commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3748,6 +3749,340 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+    size_t noteCounter = 0;
+    bool maxedOutUTXOsFlag = false;
+    bool maxedOutNotesFlag = false;
+    size_t mempoolLimit = (nUTXOLimit != 0) ? nUTXOLimit : (size_t)GetArg("-mempooltxinputlimit", 0);
```
1. Seems wrong name for this variable - it can be the tx limit user wanted unrelated to mempool.
2. Wondering if will be computed wrongly when user didn't supply `-mempooltxinputlimit` in zcashd

str4d replied:
1. Variable name was copied from `z_shieldcoinbase`. I could rename it, but feels like a cleanup.

2. If `-mempooltxinputlimit` isn't supplied, the default is set to zero, meaning "try include as many UTXOs as possible, subject to the transaction size limit".

arielgabizon replied:
I see. It's a bit confusing that "0" means no limit rather than a limit of zero.  But I can see that being useful. I would document that somehow though.

bitcartel commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3748,6 +3749,340 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+            "                             - \"ANY_ZADDR\": Merge notes from all z-addrs belonging to the wallet.\n"
+            "                         If a special string is given, any given addresses of that type will be ignored.\n"
+            "    [\n"
+            "      \"address\"            (string) Can be a t-addr or a z-addr\n"
```
Alignment. Move `(string)...` to  the left.  Two characters.

bitcartel commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3748,6 +3749,340 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+    CMutableTransaction contextualTx = CreateNewContextualCMutableTransaction(
+        Params().GetConsensus(),
+        chainActive.Height() + 1);
+
```
If v1 and any zddrs are involved as input or output, change to v2.  Code is already in z_sendmany PR https://github.com/zcash/zcash/pull/2978

bitcartel commented on src/wallet/asyncrpcoperation_mergetoaddress.cpp:
```diff
@@ -0,0 +1,926 @@
+
+    // Prepare raw transaction to handle JoinSplits
+    CMutableTransaction mtx(tx_);
+    if (!mtx.fOverwintered) {
```
Remove.  Rationale is that the caller in rpcwallet instantiates this async operation passing in a mutable tx which is the template to build on.  Format and contextual data 
 e.g. nVersion (and versionGroupId etc) are set by the caller.

bitcartel commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3748,6 +3749,340 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+            FormatMoney(mergedValue), FormatMoney(nFee)));
+    }
+
+    // Check that the user specified fee is sane (if too high, it can result in error -25 absurd fee)
```
z_mergetoaddress created a transaction with 8 joinsplits... when it finished, it triggered a mempool rejection due to absurd fee.

Command line:
```
./zcash-cli z_mergetoaddress '["zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i", "ztn8nEZgVxNGLHjC7xhCAdjBh4DC8fSWVND6GtGrhpSUAtoA6MbSuTeeo8eSCK9BcDvuU3nUuKWxGM6RGzfCSBoDK6Fdag8"]' ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 3 3
{
  "remainingUTXOs": 0,
  "remainingTransparentValue": 0.00000000,
  "remainingNotes": 2,
  "remainingShieldedValue": 3.12480000,
  "mergingUTXOs": 0,
  "mergingTransparentValue": 0.00000000,
  "mergingNotes": 10,
  "mergingShieldedValue": 3045.74586597,
  "opid": "opid-20af324c-986d-42e3-b243-8db25df5b3d1"
}
```
z_getoperationstatus returned:
```
  {
    "id": "opid-20af324c-986d-42e3-b243-8db25df5b3d1",
    "status": "failed",
    "creation_time": 1519283425,
    "error": {
      "code": -25,
      "message": ""
    },
    "method": "z_mergetoaddress",
    "params": {
      "fromaddresses": [
        "zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i",
        "ztn8nEZgVxNGLHjC7xhCAdjBh4DC8fSWVND6GtGrhpSUAtoA6MbSuTeeo8eSCK9BcDvuU3nUuKWxGM6RGzfCSBoDK6Fdag8"
      ],
      "toaddress": "ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE",
      "fee": 3.00000000
    }
  }
```
and debug.log
```
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending 3045.74586597 to send 3042.74586597 with fee 3.00
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: transparent input: 0.00
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: private input: 3045.74586597
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: private output: 3042.74586597
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: fee: 3.00
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=da408a2512, vjoinsplit=2, ciphertext=0, amount=2560.99686597, height=662, confirmations=15)
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=fdee405f5e, vjoinsplit=0, ciphertext=0, amount=1.5624, height=670, confirmations=7)
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2562.55926597)
2018-02-22 07:36:24 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 0 (vpub_old=0.00, vpub_new=0.00, in[0]=2560.99686597, in[1]=1.5624, out[0]=0.00, out[1]=2562.55926597)
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=0, n=1, zaddr=ztqRjKVqZuyBS4dBxce1qLNmbRs2DaRFMtRFu7dfgcmLVVZeH7HhgciWWkmTE7CE65EXEyVut3YjYDG8nCP7FheKjsunHJS
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=0, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2562.55926597)
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=50d6136c79, vjoinsplit=0, ciphertext=1, amount=1.5624, height=667, confirmations=10)
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2564.12166597)
2018-02-22 07:37:01 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 1 (vpub_old=0.00, vpub_new=0.00, in[0]=2562.55926597, in[1]=1.5624, out[0]=0.00, out[1]=2564.12166597)
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=1, n=1, zaddr=ztntUYUSMPS7SxYr3TWkzsJdJWJeLkX1Rdgdz1hC66678otpLdHRpd8zFQjgoERd2CqS1wv25qbPLMFFdgjLgF5UiECBaCW
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=1, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2564.12166597)
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=65822fd89e, vjoinsplit=0, ciphertext=0, amount=1.5624, height=665, confirmations=12)
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2565.68406597)
2018-02-22 07:37:38 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 2 (vpub_old=0.00, vpub_new=0.00, in[0]=2564.12166597, in[1]=1.5624, out[0]=0.00, out[1]=2565.68406597)
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=2, n=1, zaddr=ztmpsYXY9TVQnuMDyChYcoxmKo2BPoZaDnTuWgVniY3Jz15zW7VF2tCtQJQubTDmUeTzQYBhCtsoWVnvFiNPv7dUFEzX2Wp
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=2, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2565.68406597)
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=9350300fe4, vjoinsplit=0, ciphertext=1, amount=237.8748, height=673, confirmations=4)
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2803.55886597)
2018-02-22 07:38:18 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 3 (vpub_old=0.00, vpub_new=0.00, in[0]=2565.68406597, in[1]=237.8748, out[0]=0.00, out[1]=2803.55886597)
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=3, n=1, zaddr=ztkMUzdEqdejnfvfWkCaMkQNGKkGrffuodLA7Y6XjopFKFr4uY67b1gVeypQTH6y75pdBHrFn5r8gatcoVwY2nsUV35jWuh
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=3, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2803.55886597)
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=1770d39e76, vjoinsplit=0, ciphertext=1, amount=1.5624, height=669, confirmations=8)
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2805.12126597)
2018-02-22 07:38:57 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 4 (vpub_old=0.00, vpub_new=0.00, in[0]=2803.55886597, in[1]=1.5624, out[0]=0.00, out[1]=2805.12126597)
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=4, n=0, zaddr=ztN5FRhAriHE9bmxfeVx2AuhFPQM9TK3fiLMDD6fRbF9cu1319Mm4szEFiHYraSpSy6XXw7bKWHSensZvYXTswXfFcwkLJM
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=4, n=1, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2805.12126597)
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=7e093d2378, vjoinsplit=0, ciphertext=1, amount=1.5624, height=666, confirmations=11)
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2806.68366597)
2018-02-22 07:39:34 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 5 (vpub_old=0.00, vpub_new=0.00, in[0]=2805.12126597, in[1]=1.5624, out[0]=0.00, out[1]=2806.68366597)
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=5, n=1, zaddr=ztP4FsgajHhwas86iLsd4vsBCBkuhzdAMQXZT3oMwc835UMRwdjMrsyY3KD3sbRnKgwpmRxPpbkK1uhsKht1TDAdEsCUS2j
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=5, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2806.68366597)
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=8d3a2e58dc, vjoinsplit=0, ciphertext=1, amount=1.5624, height=668, confirmations=9)
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=2808.24606597)
2018-02-22 07:40:11 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 6 (vpub_old=0.00, vpub_new=0.00, in[0]=2806.68366597, in[1]=1.5624, out[0]=0.00, out[1]=2808.24606597)
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=6, n=1, zaddr=ztTkSnFnan48jUvre4qkABFGUSkzswXxcEkHbhsBaAEe87imo4yuhpHkQNd8hRnFESPGDm9cHxYgRbH3eKtUWS39KiRMhAn
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=6, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=2808.24606597)
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=3d42639d4d, vjoinsplit=0, ciphertext=0, amount=235.9374, height=662, confirmations=15)
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for change (amount=3044.18346597)
2018-02-22 07:40:49 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 7 (vpub_old=0.00, vpub_new=0.00, in[0]=2808.24606597, in[1]=235.9374, out[0]=0.00, out[1]=3044.18346597)
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=7, n=1, zaddr=ztK5XQzh4ggA3vaaDeQoxXmLiGA8M2sXAXDonC72dtv7iP8zHkqdqEJ3ThDc6iCFA7pV3UWPGdgUDo4fCHtqT9S3FFuidb9
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=7, n=0, zaddr=zth7VXAVscpJyAFeadPzuktSm1ZrbSCwLjjv5ymeqipZ1d7GfebKXEzdmWP1TutyaDnekteDb7xT8J7cSskajcEKrXKDk4i
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending change (amount=3044.18346597)
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: spending note (txid=62015dfca2, vjoinsplit=0, ciphertext=0, amount=1.5624, height=672, confirmations=5)
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: generating note for target (amount=3042.74586597)
2018-02-22 07:41:26 opid-20af324c-986d-42e3-b243-8db25df5b3d1: creating joinsplit at index 8 (vpub_old=0.00, vpub_new=3.00, in[0]=3044.18346597, in[1]=1.5624, out[0]=0.00, out[1]=3042.74586597)
2018-02-22 07:42:04 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=8, n=1, zaddr=ztKqq3VXPFQnk3fUTGjq9YcnovGnHyNMLBJgPbZpV7En3xqEdzKiUZfVFVBmEoHnTeo4S7Kykt6nwTdvXyQj6JyZJbAc2ym
2018-02-22 07:42:04 opid-20af324c-986d-42e3-b243-8db25df5b3d1: Payment Disclosure: js=8, n=0, zaddr=ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE
2018-02-22 07:42:04 ERROR: AcceptToMemoryPool: absurdly high fees 3c581d17a73c8ab675b0e128db3d4140f5d14c3258e9ac3f1a53f09823e01979, 300000000 > 16320000
2018-02-22 07:42:04 opid-20af324c-986d-42e3-b243-8db25df5b3d1: z_mergetoaddress finished (status=failed, error=)
```

str4d replied:
Not sure what to say there: you set a fee of 3 ZEC in the CLI, which was larger than what the mempool calculated as a maximum reasonable fee. `z_mergetoaddress` only checks beforehand that the fee is not greater than the value being merged, for which an absurd fee would of course be triggered, but I don't think it's necessary to duplicate the absurd fee logic everywhere else.

bitcartel replied:
Actually the fee of 3 was a typo :-)

There is a general point here where a user can't predict how big the transaction will be and what an appropriate fee is.   Also, in this call, merging 3045 ZEC for a fee of 3 ZEC, which is 0.1%, might be reasonable to some people, especially if they are sending to someone, rather than merging just for themselves. 

Here's the mempool check.

```
        if (fRejectAbsurdFee && nFees > ::minRelayTxFee.GetFee(nSize) * 10000)
            return error("AcceptToMemoryPool: absurdly high fees %s, %d > %d",
                         hash.ToString(),
                         nFees, ::minRelayTxFee.GetFee(nSize) * 10000);
```

It would be nice if we could add another check before spending a few minutes creating multiple joinsplits.  Since we know the mempool check is based on size, and we already compute `estimatedTxSize`, the following patch works and has been tested.

```
    if (nFee > netAmount) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Fee %s is greater than the net amount to be shielded %s", FormatMoney(nFee), FormatMoney(netAmount)));
    }

    CAmount estimatedFee = minRelayTxFee.GetFee(estimatedTxSize);
    CAmount estimatedAbsurdLimit = estimatedFee * 10000; // Matches code in AcceptToMemoryPool
    if (nFee > estimatedAbsurdLimit) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Fee %s is much bigger than the estimated fee %s", FormatMoney(nFee), FormatMoney(estimatedFee)));
    }
```

Here is some output with debug statements.

```
./zcash-cli z_mergetoaddress '["*"]' ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 3
nfee = 300000000
netAmount = 305387036597
estimatedTxSize = 3804
estimatedFee = 380
estimatedAbsurd = 3800000
error code: -8
error message:
Fee 3.00 is much bigger than the estimated fee 0.0000038

./zcash-cli z_mergetoaddress '["*"]' ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 0.04
nfee = 4000000
netAmount = 305683036597
estimatedTxSize = 3804
estimatedFee = 380
estimatedAbsurd = 3800000
error code: -8
error message:
Fee 0.04 is much bigger than the estimated fee 0.0000038

z_mergetoaddress '["*"]' ztJqA2UeXP5ZQWiuK8HRijfj3V7GxZNTXQkjf4WXc8Sj2Evgyen6DmeW5yhhZjTzFXhsqdS6bqXjV7oT9zu9GdzBqB6viRE 0.038
nfee = 3800000
netAmount = 305683336597
estimatedTxSize = 3804
estimatedFee = 380
estimatedAbsurd = 3800000
{
  "remainingUTXOs": 0,
  "remainingTransparentValue": 0.00000000,
  "remainingNotes": 0,
  "remainingShieldedValue": 0.00000000,
  "mergingUTXOs": 0,
  "mergingTransparentValue": 0.00000000,
  "mergingNotes": 2,
  "mergingShieldedValue": 3056.87036597,
  "opid": "opid-e9d944ae-7ad0-48c2-bfdb-886ce393cd2d"
}
```

str4d replied:
Is this something we need to fix right now? Not to appeal to authority, but I copied this fee logic from your implementation of `z_shieldcoinbase` :stuck_out_tongue_winking_eye: I would prefer that the RPC is consistent, so if we make this kind of change, we should make it in a separate PR to all our `z_*` spending RPCs.

bitcartel replied:
ticket filed

str4d commented:
> Did you find the source of the anchor issue? I haven't seen it again, so it may have been resolved with your change address fix.

Yes I did. Here's the commit I made to fix it (before squashing the PR):
```patch
commit 022f844cfa90fb4d2534b2ca44d7c05a2741f756
Author: Jack Grigg <jack@z.cash>
Date:   Fri Feb 16 18:33:10 2018 +0000

    Don't treat vpub_old as change
    
    Fixes the JoinSplit anchor bug, where the algorithm was looking for a change
    note that did not exist.

diff --git a/src/wallet/asyncrpcoperation_mergetoaddress.cpp b/src/wallet/asyncrpcoperation_mergetoaddress.cpp
index 85c3e02..de3c546 100644
--- a/src/wallet/asyncrpcoperation_mergetoaddress.cpp
+++ b/src/wallet/asyncrpcoperation_mergetoaddress.cpp
@@ -382,7 +382,6 @@ bool AsyncRPCOperation_mergetoaddress::main_impl()
             }
             info.vpub_old += vpubOldTarget; // funds flowing from public pool
             vpubOldProcessed = true;
-            jsChange += vpubOldTarget;
         }
 
         CAmount jsInputValue = 0;
@@ -548,7 +547,7 @@ bool AsyncRPCOperation_mergetoaddress::main_impl()
         }
 
         // Accumulate change
-        jsChange = jsInputValue;
+        jsChange = jsInputValue + info.vpub_old;
 
         // Set vpub_new in the last joinsplit (when there are no more notes to spend)
         if (zInputsDeque.empty()) {
```

str4d commented:
Force-pushed to address @bitcartel's comments. Here is the diff:

```diff
diff --git a/src/wallet/asyncrpcoperation_mergetoaddress.cpp b/src/wallet/asyncrpcoperation_mergetoaddress.cpp
index 823f289..5257cbc 100644
--- a/src/wallet/asyncrpcoperation_mergetoaddress.cpp
+++ b/src/wallet/asyncrpcoperation_mergetoaddress.cpp
@@ -280,9 +280,6 @@ bool AsyncRPCOperation_mergetoaddress::main_impl()
 
     // Prepare raw transaction to handle JoinSplits
     CMutableTransaction mtx(tx_);
-    if (!mtx.fOverwintered) {
-        mtx.nVersion = 2;
-    }
     crypto_sign_keypair(joinSplitPubKey_.begin(), joinSplitPrivKey_);
     mtx.joinSplitPubKey = joinSplitPubKey_;
     tx_ = CTransaction(mtx);
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index 201b493..d5769aa 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -3786,7 +3786,7 @@ UniValue z_mergetoaddress(const UniValue& params, bool fHelp)
             "                             - \"ANY_ZADDR\": Merge notes from all z-addrs belonging to the wallet.\n"
             "                         If a special string is given, any given addresses of that type will be ignored.\n"
             "    [\n"
-            "      \"address\"            (string) Can be a t-addr or a z-addr\n"
+            "      \"address\"          (string) Can be a t-addr or a z-addr\n"
             "      ,...\n"
             "    ]\n"
             "2. \"toaddress\"           (string, required) The t-addr or z-addr to send the funds to.\n"
@@ -4060,6 +4060,10 @@ UniValue z_mergetoaddress(const UniValue& params, bool fHelp)
     CMutableTransaction contextualTx = CreateNewContextualCMutableTransaction(
         Params().GetConsensus(),
         chainActive.Height() + 1);
+    bool isShielded = numNotes > 0 || isToZaddr;
+    if (contextualTx.nVersion == 1 && isShielded) {
+        contextualTx.nVersion = 2; // Tx format should support vjoinsplit
+    }
 
     // Create operation and add to global queue
     std::shared_ptr<AsyncRPCQueue> q = getAsyncRPCQueue();
```

arielgabizon commented on src/test/rpc_wallet_tests.cpp:
```diff
@@ -1409,4 +1410,312 @@ BOOST_AUTO_TEST_CASE(rpc_z_shieldcoinbase_internals)
+{
+    SelectParams(CBaseChainParams::TESTNET);
+
+    LOCK(pwalletMain->cs_wallet);
```
https://github.com/zcash/zcash/issues/2979

arielgabizon commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3748,6 +3749,344 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+    size_t numNotes = noteInputs.size();
+
+    if (numUtxos == 0 && numNotes == 0) {
+        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Could not find any funds to merge.");
```
should we add "(non-coinbase)" to clarify

str4d replied:
It's already documented in the help text, but this will also help.

arielgabizon commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3748,6 +3749,344 @@ UniValue z_shieldcoinbase(const UniValue& params, bool fHelp)
+            FormatMoney(mergedValue), FormatMoney(nFee)));
+    }
+
+    // Check that the user specified fee is sane (if too high, it can result in error -25 absurd fee)
```
You are not exactly checking that the fee is sane as the comment says- the fee maybe sane but netAmount very small,
and the fee may be insane but nAmount larger.

arielgabizon replied:
Actually, if you are relying anyway on error -25 to catch and crash, I'm not sure what you gain by this check.

str4d replied:
> the fee maybe sane but netAmount very small

I dispute that anyone would consider the fee to be sane if the fee is larger than the amount you are sending. If there is a genuine use-case for it, in the context of merging UTXOs, then we can address it later. For now, users should just use a smaller fee if the UTXO set they are merging has values so small that their total sum is less than 0.0002 ZEC.

arielgabizon replied:
Would consider adding to the comment that your goal is to catch some of the bad fees before doing the joinsplits as @bitcartel said in discussion above; otherwise it can be unclear why add this check.

str4d commented:
- Two builders passed
- One builder timed out in the new RPC test (transient failure)
- One builder failed the new RPC test (a sporadic failure I hoped I'd avoided)

I'll dig into the last of these and see if I can further reduce its likelihood.

str4d commented:
Found and fixed a missing `sync_all()` in the RPC test that was likely causing a race condition.

@zkbot r+

str4d commented:
Missed a couple of sync spots. Force-pushed the final commit to squash them in.

bitcartel commented:
@zkbot r+

arielgabizon commented on src/wallet/asyncrpcoperation_mergetoaddress.cpp:
```diff
@@ -0,0 +1,926 @@
+        std::vector<JSOutPoint> vOutPoints;
+        std::vector<boost::optional<ZCIncrementalWitness>> vInputWitnesses;
+        uint256 inputAnchor;
+        int numInputsNeeded = (jsChange > 0) ? 1 : 0;
```
Yeah, just noticed that too :), it's also in `z_sendmany`
It seems like currentInput is a more correct name.
But on that note, I don't understand the  logic below.
Operator precedence http://en.cppreference.com/w/cpp/language/operator_precedence says the ++ happens before the comparison, in the line below.
So seems like if you had a change note, you'll skip adding another input note (cause `numInputsNeeded` will be 2 on first check)
even though you want to add a second one.

arielgabizon commented on src/wallet/asyncrpcoperation_mergetoaddress.cpp:
```diff
@@ -0,0 +1,923 @@
+        // Consume spendable non-change notes
+        //
+        std::vector<Note> vInputNotes;
+        std::vector<SpendingKey> vInputZKeys;
```
What is the role of `vInputZKeys`? Why do they appear here and not in `z_sendmany`?

str4d replied:
In `z_sendmany` all the notes you spend come from the same address, so you can just have a class variable store the one spending key you need. Here, every note could be from a different address, so you need their spending keys individually.

str4d commented on src/wallet/asyncrpcoperation_mergetoaddress.cpp:
```diff
@@ -0,0 +1,926 @@
+
+        //
+        // Consume change as the first input of the JoinSplit.
+        //
```
If all you want to do is *transfer* all funds from the input addresses to the given output, then yes you don't need change outputs. But the purpose of this method is to *merge* all funds to the given output, so you need to use change outputs to do that.

arielgabizon replied:
Hmm, maybe it's just that the word change here is misleading?
iirc you use each joinsplit to unify another input note with the current accumulated output note.
And you're using the name "change" for this current accumulated note..maybe you can get used to that name.
